pkg/controller/usersignup/approval.go

```diff
@@ -0,0 +1,120 @@
+package usersignup
+
+import (
+	"context"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	crtCfg "github.com/codeready-toolchain/host-operator/pkg/configuration"
+	"github.com/codeready-toolchain/host-operator/pkg/counter"
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// getClusterIfApproved checks if the user can be approved and provisioned to any member cluster.
+// It the user can be approved then the function returns true as the first returned value, the second value contains a cluster name the user should be provisioned to.
+// If there is no suitable member cluster, then it returns empty string.
+//
+// If the user is approved manually then it tries to get member cluster with enough capacity if the target cluster is not already specified for UserSignup.
+// If the user is not approved manually, then it loads HostOperatorConfig to check if automatic approval is enabled or not. If it is then it checks
+// capacity thresholds and the actual use if there is any suitable member cluster.
+func getClusterIfApproved(cl client.Client, crtConfig *crtCfg.Config, userSignup *toolchainv1alpha1.UserSignup, getMemberClusters cluster.GetMemberClustersFunc) (bool, string, error) {
+	config := &toolchainv1alpha1.HostOperatorConfig{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: userSignup.Namespace, Name: "config"}, config); err != nil {
```

Comment:
Well.. I guess we don't want to explicitly load the host operator config in every place where we need the configuration.
Can we abstract access to the config?
Something like `config.HostOperatorConfig()` function somewhere. And reuse this function everywhere we need access to the config. The function for now could load the corresponding "config" resource but later we can convert it to a controller which will watch the "config" resource and update the internal cache if it's changed. Not sure if we really need such cache though. Maybe the go client cache is already good enough. But we can think about that cache later.

Reply:
Yes yes yes, I had the very same concern when I was writing the code. 
Actually, I started working on a controller that does exactly the same what you proposed with the cache :smile:, I just want to submit it as a separated PR and when this automatic approval is done. See this: https://github.com/MatousJobanek/host-operator/tree/hostoperatorconfig-controller/pkg/controller/hostoperatorconfig
For now, I decided to load it here and then replace it with calling the cache.

Reply:
OK. Makes sense. Let's keep it for a separate PR then!